src/mc-fle2-rfds.c

```diff
@@ -0,0 +1,629 @@
+      bson_t elem;
+      TRY (BSON_APPEND_DOCUMENT_BEGIN (&and, "0", &elem));
+      bson_t operator;
+      TRY (BSON_APPEND_ARRAY_BEGIN (
+         &elem, mc_FLE2RangeOperator_to_string (spec->firstOp), &operator));
+      TRY (BSON_APPEND_UTF8 (&operator, "0", spec->field));
+      TRY (_mongocrypt_buffer_append (&p1, &operator, "1", 1));
+      TRY (bson_append_array_end (&elem, &operator));
+      TRY (bson_append_document_end (&and, &elem));
+
+      if (spec->nOps == 2) {
+         TRY (BSON_APPEND_DOCUMENT_BEGIN (&and, "1", &elem));
+         TRY (BSON_APPEND_ARRAY_BEGIN (
+            &elem, mc_FLE2RangeOperator_to_string (spec->secondOp), &operator));
+         TRY (BSON_APPEND_UTF8 (&operator, "0", spec->field));
+         TRY (_mongocrypt_buffer_append (&p2, &operator, "1", 1));
+         TRY (bson_append_array_end (&elem, &operator));
+         TRY (bson_append_document_end (&and, &elem));
+      }
+
+      TRY (bson_append_array_end (out, &and));
+   } else {
+      /*
+      Create a Match Expression document like:
+      {
+        "$and" : [
+            {"age" : { "$gt": "<placeholder1>" }},
+            {"age" : { "$lt": "<placeholder2>" }}
+         ]
+      }
+      */
+      bson_t and;
+      TRY (BSON_APPEND_ARRAY_BEGIN (out, "$and", &and));
+      bson_t elem;
+      TRY (BSON_APPEND_DOCUMENT_BEGIN (&and, "0", &elem));
+      bson_t operator;
+      TRY (BSON_APPEND_DOCUMENT_BEGIN (&elem, spec->field, &operator));
+      const char *op_str = mc_FLE2RangeOperator_to_string (spec->firstOp);
+      TRY (_mongocrypt_buffer_append (&p1, &operator, op_str, -1));
+      TRY (bson_append_document_end (&elem, &operator));
+      TRY (bson_append_document_end (&and, &elem));
+
+      if (spec->nOps == 2) {
+         TRY (BSON_APPEND_DOCUMENT_BEGIN (&and, "1", &elem));
+         TRY (BSON_APPEND_DOCUMENT_BEGIN (&elem, spec->field, &operator));
+         const char *op_str = mc_FLE2RangeOperator_to_string (spec->secondOp);
+         TRY (_mongocrypt_buffer_append (&p2, &operator, op_str, -1));
+         TRY (bson_append_document_end (&elem, &operator));
+         TRY (bson_append_document_end (&and, &elem));
+      }
+
+      TRY (bson_append_array_end (out, &and));
+   }
+
+#undef TRY
+   ok = true;
+fail:
+   _mongocrypt_buffer_cleanup (&p2);
+   _mongocrypt_buffer_cleanup (&p1);
+   bson_destroy (&infDoc);
+   return ok;
+}
+
+static mlib_once_flag payloadId_init_flag = MLIB_ONCE_INITIALIZER;
+static mongocrypt_mutex_t payloadId_mutex;
+static int32_t payloadId = 0;
+
+static void
+payloadId_init_mutex (void)
+{
+   _mongocrypt_mutex_init (&payloadId_mutex);
+}
+
+void
+mc_reset_payloadId_for_testing (void)
+{
+   mlib_call_once (&payloadId_init_flag, payloadId_init_mutex);
+   MONGOCRYPT_WITH_MUTEX (payloadId_mutex)
+   {
+      payloadId = 0;
+   }
+}
+
+// mc_getNextPayloadId is thread safe.
+int32_t
+mc_getNextPayloadId (void)
+{
+   mlib_call_once (&payloadId_init_flag, payloadId_init_mutex);
+   int32_t ret;
```

Comment:
Is it likely that we will overflow the int32? Would it be a problem to have duplicate IDs for a very-long-running process?

Reply:
No. The payloadId field is used in [server-side validation](https://docs.google.com/document/d/1JLrpDNP1wjJCj-9dkQN9wbvSMxKe9MqL0XNF8dFbMuk/edit#heading=h.hemu885zhbfu) for one command:

> Every comparison operator with an encrypted range payload and a defined secondOp must have a sibling predicate with the same predicate ID and of the opposite sign ($gt and $lte, or $gt and $lt, for example).

The requirement is payloads within one command have distinct payloadIds.